# Quantile forecasting

## 1. Data Preparation

This section loads data with `setup()` and prepares features/targets for quantile forecasting as a **risk-envelope tool**.


### Installation

In [1]:
!pip install scikit-learn pandas numpy joblib


[notice] A new release of pip is available: 25.0.1 -> 26.0.1
[notice] To update, run: python.exe -m pip install --upgrade pip


### Importation of ```setup()``` method

In [2]:
import sys
from pathlib import Path

sys.path.append(str(Path("../../src").resolve()))

from setup import setup

### Application of the method

I start by loading the dataset using my ```setup()``` function.

In [3]:
data = setup(
    csv_path="../../data/processed/financial_tools_datset.csv",
    target_col="Price",
    horizon=1
)

### Training and valuation splits

Here, I’m basically transforming raw market data into a supervised learning problem. I define my target as the future return, not the price itself. That’s important, because I care about movement, not absolute value.

I also split the data in time order, not randomly. If I shuffled this, I would leak future information into the past and completely destroy the validity of the model.

Finally, I normalize the features using StandardScaler. I do this so the model doesn’t get biased by scale differences across features.

👉 At this point, I have a clean mapping:

>Inputs → market state\
Output → future return

In [4]:
X_train = data["X_train"]
X_val   = data["X_val"]
X_test  = data["X_test"]

y_train = data["y_train"]
y_val   = data["y_val"]
y_test  = data["y_test"]

## The model + gradient boost

Now I train models to predict quantiles, not averages.

Instead of asking:

> “What will the return be?”

I’m asking:

> “What is a realistic lower bound, median, and upper bound for the return?”

Each model is trained with a different alpha:

0.1 → pessimistic scenario\
0.5 → median\
0.9 → optimistic scenario

So instead of a single prediction, I’m building a range of possible outcomes.

👉 This is the first step toward modeling uncertainty explicitly.

In [5]:
from sklearn.ensemble import GradientBoostingRegressor

quantiles = [0.1, 0.5, 0.9]
models = {}

for q in quantiles:
    model = GradientBoostingRegressor(
        loss="quantile",
        alpha=q,
        n_estimators=200,
        max_depth=3,
        learning_rate=0.05
    )
    
    model.fit(X_train, y_train)
    models[q] = model

Here I simply run each trained model on the test set. Each model outputs a different view of the future:

1. One says “things could go this bad”

2. Another says “this is the most typical outcome”

Another says “this is the best-case scenario”

I store all of them separately because they each represent a different slice of the distribution.\
👉 I’m no longer dealing with a point estimate — I now have a distribution approximation.

In [6]:
preds = {}

for q, model in models.items():
    preds[q] = model.predict(X_test)

q10 = preds[0.1]
q50 = preds[0.5]
q90 = preds[0.9]

### Avaliation (Pinball Loss)

I evaluate each quantile using Pinball Loss. This is important: I cannot use MSE or anything like that. Those assume I’m predicting a mean. I’m not.

Each quantile has its own loss, because each one represents a different objective:

- Lower quantile → penalizes underestimation differently
- Upper quantile → penalizes overestimation differently

>👉 If I evaluate this the wrong way, I’ll think the model is good when it’s actually garbage.

In [7]:
import numpy as np

def pinball_loss(y_true, y_pred, q):
    return np.mean(np.maximum(q*(y_true - y_pred), (q-1)*(y_true - y_pred)))

for q in quantiles:
    loss = pinball_loss(y_test, preds[q], q)
    print(f"Quantile {q}: {loss:.6f}")

Quantile 0.1: 0.000696
Quantile 0.5: 0.001230
Quantile 0.9: 0.000906


What this tells me

- The values are low and consistent → no obvious instability
- q0.5 (median) has the highest loss → expected
    - It’s the hardest quantile to fit in noisy financial data
- q0.1 and q0.9 are lower → tails are being fit “easier”

### Interpretation

Our model is not broken. There’s no immediate red flag. Scale looks reasonable for return data.
> **Variable intervals → model captured heteroscedasticity**

### Calibration

**Generally expected: 0.8**

In [8]:
coverage = ((y_test >= q10) & (y_test <= q90)).mean()
print(f"Coverage [0.1–0.9]: {coverage:.3f}")

Coverage [0.1–0.9]: 0.782


Good. This is actually solid.

**Quick read**: 
- Expected: 0.8
- Got: 0.782

>👉 Error ≈ -0.018 (very small)

### Interpretation

- Slight undercoverage → model is a bit overconfident

    - **But**: Deviation is small!

Totally acceptable for a first baseline

### Risk envelope

Now I turn predictions into something operational.

I build an interval:

- Lower bound → q10
- Upper bound → q90

Then I compute the width of this interval. This gives me:
1. How uncertain the market is right now
2. How wide my expected outcomes are

**👉 This is directly usable for:**

1. position sizing
2. stop-loss distance
3. capital allocation

This is where the model stops being “ML” and becomes trading infrastructure.

In [9]:
import pandas as pd

df_results = pd.DataFrame({
    "y_true": y_test,
    "q10": q10,
    "q50": q50,
    "q90": q90
})

df_results["interval_width"] = df_results["q90"] - df_results["q10"]
df_results.head()

,y_true,q10,q50,q90,interval_width
0,0.000461,-0.002864,0.001176,0.005967,0.008831
1,0.001199,-0.002864,-0.000300,0.004799,0.007663
2,-0.003500,-0.004679,-0.002446,0.004483,0.009162
3,-0.000370,-0.001504,-0.000060,0.004731,0.006236
4,0.000925,-0.000458,0.000563,0.004759,0.005217


# Envelope signs derivation

Definition:

- long_signal = (q10 > 0)
- short_signal = (q90 < 0)

Meaning:\
- Long signal → even the pessimistic scenario (10th percentile) is positive
- Short signal → even the optimistic scenario (90th percentile) is negative

Interpretation:\
These are high-confidence directional trades. We are filtering only situations where the entire predicted distribution is on one side of zero. We built a strict decision filter based on distribution support, not probability.

In [10]:
df_results["long_signal"] = df_results["q10"] > 0
df_results["short_signal"] = df_results["q90"] < 0
print("Long signals:", df_results["long_signal"].sum())
print("Short signals:", df_results["short_signal"].sum())

Long signals: 18
Short signals: 0


### Expectation measure

Definition:

- ```expected_value = q50```
Meaning:\
The median of the predicted conditional distribution. Robust estimate of central tendency (less sensitive than mean).\
Interpretation:
- Represents the most likely return outcome
- Serves as your baseline directional bias

In [11]:
df_results["expected_value"] = df_results["q50"]
print(df_results["expected_value"].describe())

count    206.000000
mean       0.000097
std        0.002911
min       -0.008054
25%       -0.001942
50%        0.000198
75%        0.002106
max        0.009961
Name: expected_value, dtype: float64


### Risk-adjusted signal

Definition:
- risk_reward = q50 / (q90 - q10)
Meaning:\
Signal normalized by its uncertainty (interval width).

Interpretation:
- High value → strong signal + low uncertainty
- ow value → weak signal or high uncertainty

In [12]:
df_results["risk_reward"] = df_results["q50"] / df_results["interval_width"]
print(df_results["risk_reward"].describe())

count    206.000000
mean       0.022715
std        0.317909
min       -0.743453
25%       -0.191934
50%        0.016352
75%        0.209013
max        0.893803
Name: risk_reward, dtype: float64


### Acerto direcional

In [13]:
direction = (df_results["y_true"] > 0).astype(int)
pred_direction = (df_results["q50"] > 0).astype(int)

accuracy = (direction == pred_direction).mean()
print(f"Directional accuracy: {accuracy:.3f}")

Directional accuracy: 0.801


# PnL

In [14]:
df_results["position"] = 0
df_results.loc[df_results["long_signal"], "position"] = 1
df_results.loc[df_results["short_signal"], "position"] = -1

df_results["strategy_return"] = df_results["position"] * df_results["y_true"]

print("Total return:", df_results["strategy_return"].sum())
print("Mean return:", df_results["strategy_return"].mean())
print("Sharpe (proxy):", df_results["strategy_return"].mean() / df_results["strategy_return"].std())

Total return: 0.11104754228703206
Mean return: 0.0005390657392574372
Sharpe (proxy): 0.22313784277270124


# LightGBM + more quantiles

In [15]:
!pip install lightgbm


[notice] A new release of pip is available: 25.0.1 -> 26.0.1
[notice] To update, run: python.exe -m pip install --upgrade pip


## 2. Model Construction

Train the Quantile Forecasting model (LightGBM quantile regression) to produce probabilistic outputs used by the envelope layer.


### Training the model LightGBM

In [16]:
import lightgbm as lgb

quantiles = [0.05, 0.25, 0.5, 0.75, 0.95]
models = {}

for q in quantiles:
    model = lgb.LGBMRegressor(
        objective="quantile",
        alpha=q,
        n_estimators=300,
        learning_rate=0.03,
        max_depth=5
    )
    
    model.fit(X_train, y_train)
    models[q] = model

[LightGBM] [Info] Auto-choosing col-wise multi-threading, the overhead of testing was 0.001534 seconds.
You can set `force_col_wise=true` to remove the overhead.
[LightGBM] [Info] Total Bins 5778
[LightGBM] [Info] Number of data points in the train set: 956, number of used features: 24
[LightGBM] [Info] Start training from score -0.007882
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, 

### Prediction

In [17]:
import pandas as pd

X_test_df = pd.DataFrame(X_test, columns=data["feature_columns"])

preds = {q: model.predict(X_test_df) for q, model in models.items()}

q05, q25, q50, q75, q95 = preds.values()

In [18]:
preds

{0.05: array([-0.00321114, -0.0032435 , -0.00399616, -0.00165733, -0.00191032,
        -0.00287715, -0.00125484, -0.00356045, -0.00544622, -0.00294754,
        -0.00256587, -0.00268523, -0.00121711, -0.00109143, -0.00353135,
        -0.00344693, -0.00127119, -0.00275406, -0.00297699, -0.00350093,
        -0.00479575, -0.00293711, -0.00630758, -0.00451573, -0.00890828,
        -0.0044532 , -0.005025  , -0.00382649, -0.00392513, -0.00299547,
        -0.00421791, -0.00534374, -0.00703371, -0.00448594, -0.00143   ,
        -0.00126298, -0.00456328, -0.00267027, -0.00437475, -0.00310787,
        -0.00319203, -0.00582175, -0.00519945, -0.00401251, -0.00594787,
        -0.00525446, -0.00664286, -0.00880644, -0.00686046, -0.00648034,
        -0.00637814, -0.00645804, -0.00624182, -0.00576738, -0.00573405,
        -0.00521846, -0.0059948 , -0.00874112, -0.00599708, -0.00759513,
        -0.00590376, -0.00454248, -0.00802619, -0.00701673, -0.00354508,
        -0.00303336, -0.00389854, -0.00353087

## 4. Generation of Risk Features

From quantile predictions, derive risk-envelope features (lower/upper bounds, width, and skew/asymmetry).


# Generating outputs for QF

## Final dataset of QF

In [19]:
df_qf = pd.DataFrame({
    "y_true": y_test,
    "q05": q05,
    "q25": q25,
    "q50": q50,
    "q75": q75,
    "q95": q95
})

### Width

In [20]:
df_qf["uncertainty"] = df_qf["q95"] - df_qf["q05"]

### Asymetry

In [21]:
df_qf["asymmetry"] = (df_qf["q95"] + df_qf["q05"] - 2*df_qf["q50"])

In [22]:
df_qf["quantile_lower"] = df_qf["q05"]
df_qf["quantile_upper"] = df_qf["q95"]
df_qf["quantile_width"] = df_qf["uncertainty"]
df_qf["quantile_skew"] = df_qf["asymmetry"]


In [23]:
df_qf

,y_true,q05,q25,q50,q75,q95,uncertainty,asymmetry,quantile_lower,quantile_upper,quantile_width,quantile_skew
0,0.000461,-0.003211,0.000852,0.001092,0.002609,0.005292,0.008503,-0.000104,-0.003211,0.005292,0.008503,-0.000104
1,0.001199,-0.003244,-0.000138,0.000380,0.000872,0.004420,0.007663,0.000416,-0.003244,0.004420,0.007663,0.000416
2,-0.003500,-0.003996,-0.003986,-0.003723,-0.002388,0.003642,0.007638,0.007093,-0.003996,0.003642,0.007638,0.007093
3,-0.000370,-0.001657,-0.000869,-0.000490,0.000607,0.004697,0.006354,0.004020,-0.001657,0.004697,0.006354,0.004020
4,0.000925,-0.001910,-0.000671,0.000472,0.001081,0.004777,0.006687,0.001924,-0.001910,0.004777,0.006687,0.001924
...,...,...,...,...,...,...,...,...,...,...,...,...
201,0.000354,-0.004978,-0.000988,0.000764,0.001976,0.005161,0.010139,-0.001344,-0.004978,0.005161,0.010139,-0.001344
202,0.001682,-0.004360,-0.000091,0.001486,0.002758,0.005384,0.009744,-0.001949,-0.004360,0.005384,0.009744,-0.001949
203,0.004861,-0.004589,0.000477,0.002170,0.003273,0.004472,0.009061,-0.004456,-0.004589,0.004472,0.009061,-0.004456
204,-0.006157,-0.007251,-0.006306,-0.005294,-0.003485,0.002438,0.009688,0.005775,-0.007251,0.002438,0.009688,0.005775


## 5. Output Standardization

Build a single risk-feature table aligned to the test index/order from `setup()` and ready to merge with other envelope features.


### Minimum output for RL

In [24]:
df_qf_final = df_qf[[
    "quantile_lower",
    "q50",
    "quantile_upper",
    "quantile_width",
    "quantile_skew"
]]


## 3. Sanity Checks

Confirm predictions/risk features are present, free of NaN, and have plausible distributions.


### Visual outputs

In [25]:
print("Has NaN in df_qf_final:", df_qf_final.isna().any().any())
print("Rows in risk feature table:", len(df_qf_final))


Has NaN in df_qf_final: False
Rows in risk feature table: 206


In [26]:
print(df_qf_final.describe())

       quantile_lower         q50  quantile_upper  quantile_width  \
count      206.000000  206.000000      206.000000      206.000000   
mean        -0.004918    0.000107        0.005350        0.010268   
std          0.002197    0.003322        0.002452        0.002071   
min         -0.011736   -0.009000        0.000388        0.006122   
25%         -0.006427   -0.002346        0.003428        0.008835   
50%         -0.004932    0.000108        0.005140        0.009978   
75%         -0.003165    0.002467        0.006891        0.011370   
max         -0.001091    0.010509        0.013936        0.017873   

       quantile_skew  
count     206.000000  
mean        0.000219  
std         0.003610  
min        -0.009315  
25%        -0.002428  
50%         0.000076  
75%         0.002765  
max         0.008135  


In [27]:
print("Mean uncertainty:", df_qf["uncertainty"].mean())
print("Mean asymmetry:", df_qf["asymmetry"].mean())

Mean uncertainty: 0.010268181747625036
Mean asymmetry: 0.00021875360803629053


### Minimum scope of paramaters

In [28]:
quantiles = [0.05, 0.25, 0.5, 0.75, 0.95]

model = lgb.LGBMRegressor(
    objective="quantile",
    n_estimators=300,
    learning_rate=0.03,
    max_depth=5
)

## 6. Saving the Feature

Persist the standardized Quantile Forecasting risk-feature table in Parquet format for downstream RL/envelope merges.


In [29]:
df_qf_final

,quantile_lower,q50,quantile_upper,quantile_width,quantile_skew
0,-0.003211,0.001092,0.005292,0.008503,-0.000104
1,-0.003244,0.000380,0.004420,0.007663,0.000416
2,-0.003996,-0.003723,0.003642,0.007638,0.007093
3,-0.001657,-0.000490,0.004697,0.006354,0.004020
4,-0.001910,0.000472,0.004777,0.006687,0.001924
...,...,...,...,...,...
201,-0.004978,0.000764,0.005161,0.010139,-0.001344
202,-0.004360,0.001486,0.005384,0.009744,-0.001949
203,-0.004589,0.002170,0.004472,0.009061,-0.004456
204,-0.007251,-0.005294,0.002438,0.009688,0.005775


In [30]:
df_qf_final.to_parquet("dataset_envelope_quantile_forecasting.parquet", index=True)
print("Saved: dataset_envelope_quantile_forecasting.parquet")


Saved: dataset_envelope_quantile_forecasting.parquet


The resulting file contains envelope-ready risk features generated from quantile forecasts: `quantile_lower`, `q50`, `quantile_upper`, `quantile_width`, and `quantile_skew`.
